client/src/client.rs

```diff
@@ -1043,6 +1043,27 @@ pub trait RpcApi: Sized {
         self.call("uptime", &[])
     }
 
+    /// Submit a block
+    fn submit_block(&self, block: &bitcoin::Block) -> result::Result<(), SubmitBlockError> {
+        let block_hex: String = bitcoin::consensus::encode::serialize_hex(block);
+        self.submit_block_hex(&block_hex)
+    }
+
+    /// Submit a raw block
+    fn submit_block_bytes(&self, block_bytes: &[u8]) -> result::Result<(), SubmitBlockError> {
+        let block_hex: String = block_bytes.to_hex();
+        self.submit_block_hex(&block_hex)
+    }
+
+    /// Submit a block as a hex string
+    fn submit_block_hex(&self, block_hex: &str) -> result::Result<(), SubmitBlockError> {
+        match self.call("submitblock", &[into_json(&block_hex)?]) {
+            Ok(serde_json::Value::Null) => Ok(()),
+            Ok(res) => Err(SubmitBlockError(res.to_string())),
+            Err(err) => Err(err.into()),
```

Comment:
Hmm, this is quite bad as it totally obfuscates the underlying error (it won't be possible anymore to distinguish a network error from a JSON error, f.e.).

How about we add a new variant to our `Error` type for cases like this:
```diff
diff --git a/client/src/error.rs b/client/src/error.rs
index 70676c4..49751fd 100644
--- a/client/src/error.rs
+++ b/client/src/error.rs
@@ -29,6 +29,8 @@ pub enum Error {
     InvalidCookieFile,
     /// The JSON result had an unexpected structure.
     UnexpectedStructure,
+    /// The daemon returned an error string.
+    ReturnedError(String),
 }
```

So that the `Ok(serde_json::Value::String(s))` case can be mapped to an `Err(Error::ReturnedError(s))`?

Reply:
I'm a bit confused, which case do yo to you exactly refer to? Generally wrapping `Error` in `SubmitBlockError`? 
And would the new `ReturnedError` variant be on top or instead of the `SubmitBlockError` type?

Reply:
No, removing the `SubmitBlockError` type and just returning the general `Error`. Then when the node responds with a string, convert that into an `Error::ReturnedError` variant.

Reply:
Okay, I introduced the dedicated error type to account for the idiosyncratic behavior of `submitblock`  after the discussion above and following the suggestion of @shesek. However, since I personally have no strong feelings about this I'd be perfectly happy to go with the `ReturnedError` variant in lieu of the `SubmitBlockError` type, if thats okay for everyone.

Reply:
I didn't fully consider that the method could also return other generic JSONRPC errors and not just the specialized `SubmitBlockError`. Having to convert other errors into a string to fit into it (and losing its original type and associated info) isn't great.

I would go with @stevenroose's suggestion of a new variant on the existing `Error` type.

Apologies for my earlier incorrect suggestion! >_<